content/blog/2017/2017-01-18-refactoring-conditional-to-pipeline.adoc

```diff
@@ -0,0 +1,475 @@
+                expression { return params.REQUESTED_ACTION == 'greeting' }
+            }
+            steps {
+                echo "Hello, bitwiseman!"
+            }
+        }
+    }
+}
+// Script //
+properties ([
+    parameters ([
+        choice (
+            // choices must be newline separated
+            choices: ['greeting','silence'].join('\n'),
+            description: '',
+            name : 'REQUESTED_ACTION')
+    ])
+])
+
+node {
+    stage ('Speak') {
+        // Only say hello if a "greeting" is requested
+        if (params.REQUESTED_ACTION == 'greeting') {
+            echo "Hello, bitwiseman!"
+        }
+    }
+}
+----
+
+When I run this project with `REQUESTED_ACTION` set to "greeting", here's the output:
+
+[source]
+----
+[Pipeline] node
+Running on osx_mbp in /Users/bitwiseman/jenkins/agents/osx_mbp/workspace/pipeline-conditional
+[Pipeline] {
+[Pipeline] stage
+[Pipeline] { (Speak)
+[Pipeline] echo
+Hello, bitwiseman!
+[Pipeline] }
+[Pipeline] // stage
+[Pipeline] }
+[Pipeline] // node
+[Pipeline] End of Pipeline
+Finished: SUCCESS
+----
+
+When at I pass the value "silence", the only change is "Hello, bitwiseman!" is not printed.
+
+Some might argue that the Pipeline code is a bit harder to understand on first reading.
+Others would say the UI is just as confusing if not more so.
+Either way, the Pipeline representation is considerably more compact than the Jenkins UI presentation.
+Pipeline also lets us add helpful comments, which we can't do in the Freestyle UI.
+And we can easily put this Pipeline in a `Jenkinsfile` to be code-reviewed, checked-in, and versioned
+along with the rest of our code.
+
+=== Conditions
+The previous example showed the "Strings match" condition and it's Pipeline equivalent.
+Let's look at couple more interesting conditions and their Jenkins Pipeline equivalents.
+
+==== Boolean condition
+
+You might think that a boolean condition would be the simplest condition, but it isn't.
+Since it works with string values from tokens, the Conditional BuildStep plugin offers
+a number of ways to indicate true or false.
+Truth is a case insensitive match of one of the following:
+1 (the number one), Y, YES, T, TRUE, ON or RUN.
+
+Pipeline can duplicate these, but depending on the scenario we might consider
+whether a simpler expression would suffice.
+
+.Pipeline
+[pipeline]
+----
+// Declarative //
+when {
+    // case insensitive regular expression for truthy values
+    expression { return token ==~ /(?i)(Y|YES|T|TRUE|ON|RUN)/ }
+}
+steps {
+    /* step */
+}
+
+// Script //
+// case insensitive regular expression for truthy values
+if (token ==~ /(?i)(Y|YES|T|TRUE|ON|RUN)/) {
+    /* step */
+}
+----
+
+==== Logical "OR" of conditions
+
+This condition wraps other conditions.
+It takes their results as inputs and performs a logical "or" of the results.
+The `AND` and `NOT` conditions do the same, performing their respective operations.
```

Comment:
@daniel-beck 
Just to be clear, I started to put all the conditions into this post and it got pretty long.  I decided to limit myself to a few examples instead.  I  have a task to write documentation that covers this area exhaustively - At least all the condition operations.  There should probably be a separate document for tokens as well.